invoice/src/main/java/org/killbill/billing/invoice/usage/ContiguousIntervalUsageInArrear.java

```diff
@@ -129,23 +129,21 @@ public ContiguousIntervalUsageInArrear build(final boolean closedInterval) {
         }
         final LocalDate endDate = closedInterval ? internalTenantContext.toLocalDate(billingEvents.get(billingEvents.size() - 1).getEffectiveDate()) : targetDate;
 
-        final BillingIntervalDetail bid = new BillingIntervalDetail(startDate, endDate, targetDate, getBCD(), usage.getBillingPeriod(), usage.getBillingMode());
-
-        int numberOfPeriod = 0;
-        // First billingCycleDate prior startDate
-        LocalDate nextBillCycleDate = bid.getFutureBillingDateFor(numberOfPeriod);
         if (startDate.compareTo(rawUsageStartDate) >= 0) {
             transitionTimes.add(startDate);
         }
-        while (!nextBillCycleDate.isAfter(endDate)) {
-            if (nextBillCycleDate.isAfter(startDate)) {
-                if (nextBillCycleDate.compareTo(rawUsageStartDate) >= 0) {
-                    transitionTimes.add(nextBillCycleDate);
-                }
+
+        for (int i = 0; i < billingEvents.size(); i++) {
+            final BillingEvent billingEvent = billingEvents.get(i);
+            if (i == billingEvents.size() - 1) {
+                addTransitionTimesForBillingEvent(startDate, endDate, billingEvent.getBillCycleDayLocal());
+            } else {
+                final BillingEvent nextBillingEvent = billingEvents.get(i + 1);
+                final LocalDate nextEndDate = internalTenantContext.toLocalDate(nextBillingEvent.getEffectiveDate());
+                addTransitionTimesForBillingEvent(startDate, nextEndDate, billingEvent.getBillCycleDayLocal());
```

Comment:
I am confused why we reuse the same `startDate` on each iteration though: Shouldn't the `addTransitionTimesForBillingEvent` return the new `startDate` for the next iteration? Are we not gonna end up with duplicate transitions -- or even worst incorrect transitions as the BCD could now change?

Reply:
I actually went back and forth on that one...

The way we use `BillingIntervalDetail` here is a bit odd as we only want to access `getFutureBillingDateFor(numberOfPeriod)`. `startDate` in `BillingIntervalDetail` is only used to compute the `firstBillingCycleDate`, but because we will always call `alignProposedBillCycleDate`, I don't think the value matters. The code could probably benefit from some refactoring but I wanted to minimize the changes.

That being said, I've changed it in cde63173c83f82dd10dd323536241857c31a5dc2 and added a unit test with two BCD changes.